apollo-router/src/plugins/connectors/http_json_transport.rs

```diff
@@ -77,43 +78,72 @@ pub(crate) fn make_request(
         &flat_inputs,
     )?;
 
-    let (json_body, body, apply_to_errors) = if let Some(ref selection) = transport.body {
+    let request = http::Request::builder()
+        .method(transport.method.as_str())
+        .uri(uri.as_str());
+
+    // add the headers and if content-type is specified, we'll check that when constructing the body
+    let (mut request, content_type) = add_headers(
+        request,
+        original_request.supergraph_request.headers(),
+        &transport.headers,
+        &flat_inputs,
+    );
+
+    let is_form_urlencoded = content_type
+        .as_ref()
+        .map(|v| v.to_str().unwrap_or_default())
+        == Some("application/x-www-form-urlencoded");
+
+    let (json_body, form_body, body, apply_to_errors) = if let Some(ref selection) = transport.body
+    {
         let (json_body, apply_to_errors) = selection.apply_with_vars(&json!({}), &inputs);
+        let mut form_body = None;
         let body = if let Some(json_body) = json_body.as_ref() {
-            hyper::Body::from(serde_json::to_vec(json_body)?)
+            if is_form_urlencoded {
+                let encoded = encode_json_as_form(json_body.as_object().unwrap());
```

Comment:
Is this guaranteed by some other piece of code to never not be an object? I feel like we'd want to return an error, not panic.

Reply:
correct! good catch